crates/recursor/src/recursor_dns_handle.rs

```diff
@@ -226,6 +262,82 @@ impl RecursorDnsHandle {
         }
     }
 
+    /// Handle CNAME expansion for the current query
+    #[async_recursion]
+    async fn resolve_cnames(
+        &self,
+        mut lookup: Lookup,
+        query: Query,
+        now: Instant,
+        query_has_dnssec_ok: bool,
+        mut depth: u8,
+    ) -> Result<Lookup, Error> {
+        let query_type = query.query_type();
+        let query_name = query.name().clone();
+
+        self.record_cache
+            .insert_records(query, lookup.records().iter().map(Record::to_owned), now);
+
+        // Don't resolve CNAME lookups for a CNAME (or ANY) query
+        if query_type == RecordType::CNAME || query_type == RecordType::ANY {
+            return Ok(lookup);
+        }
+
+        let mut cname_chain = vec![];
+        for cname in lookup
+            .records()
+            .iter()
+            .filter(|record| record.record_type() == RecordType::CNAME)
```

Comment:
I'd just do this filter inside the `for`-loop, should result in formatting that's easier to follow and there's not much benefit to doing it this way. Might even be able to somehow fold the `record_type() == RecordType::CNAME` and `if let CNAME(name) = cname.data()` with `try_borrow()` or something?